internal/graph/check_fast_path.go

```diff
@@ -994,10 +985,119 @@ func (c *LocalChecker) recursiveTTUFastPath(ctx context.Context, req *ResolveChe
+}
+
+// objectProvider is an interface that abstracts the building of a channel that holds object IDs or usersets.
+type objectProvider interface {
+	close()
+	build(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error)
+}
+
+type simpleRecursiveObjectProvider struct {
+	mapper storage.TupleMapper
+}
+
+var _ objectProvider = &simpleRecursiveObjectProvider{}
+
+func (s *simpleRecursiveObjectProvider) close() {
+	if s.mapper != nil {
+		s.mapper.Stop()
+	}
+}
+
+func (s *simpleRecursiveObjectProvider) build(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	ds, _ := storage.RelationshipTupleReaderFromContext(ctx)
+
+	// Note: we set sortContextualTuples to false because we don't care about ordering of results,
+	// since the consumer is using hashsets to check for intersection.
+	userIter, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req,
+		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+		nil, false)
+	if err != nil {
+		return nil, err
+	}
+	usersetFromUserIter := storage.WrapIterator(storage.ObjectIDKind, userIter)
+	s.mapper = usersetFromUserIter
+	userToUsersetMessageChan := streamedLookupUsersetFromIterator(ctx, usersetFromUserIter)
+	return userToUsersetMessageChan, nil
+}
+
+type complexRecursiveObjectProvider struct {
+}
+
+var _ objectProvider = &complexRecursiveObjectProvider{}
+
+func (c *complexRecursiveObjectProvider) close() {
+}
+
+func (c *complexRecursiveObjectProvider) build(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
+	objectType, relation := tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()
+	userType := tuple.GetType(req.GetTupleKey().GetUser())
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	_, operands := typesys.IsRelationWithRecursiveTTUAndAlgebraicOperations(objectType, relation, userType)
+
+	outChannel := make(chan usersetMessage, 1)
+	defer close(outChannel)
+	pool := concurrency.NewPool(ctx, len(operands))
+
+	for _, operand := range operands {
+		pool.Go(func(ctx context.Context) error {
+			newReq := req.clone()
+			newReq.TupleKey.Relation = operand.RelationName
+			chanIterator, err := fastPathRewrite(ctx, newReq, operand.Rewrite)
+			if err != nil {
+				return err
+			}
+			for iteratorMessage := range chanIterator {
+				if iteratorMessage.err != nil {
+					return iteratorMessage.err
+				}
+				for {
+					t, err := iteratorMessage.iter.Next(ctx)
+					if err != nil {
+						iteratorMessage.iter.Stop()
+						if storage.IterIsDoneOrCancelled(err) {
+							break
+						}
+						return err
+					}
+					userset := t.GetObject()
+					concurrency.TrySendThroughChannel(ctx, usersetMessage{userset: userset}, outChannel)
+				}
+			}
+			return nil
+		})
+	}
+
+	err := pool.Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	return outChannel, nil
```

Comment:
So by the time the channel is yielded, all the goroutines have been processed? That doesn't allow you to wrap up early on a match (or a ctx cancellation).

Reply:
Fixed